sdks/python/apache_beam/yaml/yaml_join.py

```diff
@@ -0,0 +1,161 @@
+  if not isinstance(equalities, list):
+    return error
+  input_edge_list = []
+  for equality in equalities:
+    if len(equality) != 2 and not isinstance(equality, dict):
+      return error
+
+    for input, col in equality.items():
+      # TODO: look for an easier way to get field names in a Pcollection obj
+      possible_cols = set()
+      for field in pcoll['f'].element_type._fields:
+        possible_cols.add(field[0])
+      if input not in pcoll.keys() or col not in possible_cols:
+        return ValueError(
+          'Invalid input alias or column name doesn\'t exist in the input')
+
+    input_edge_list.append(tuple(equality.keys()))
+
+  if not nx.is_connected(nx.Graph(input_edge_list)):
+    return ValueError(
+      'Inputs in equalities are not all connected'
+    )
+
+
+def _parse_fields(tables, fields):
+  # TODO(titodo) - consider taking all validations to a preprocessing fn
+  output_fields = []
+  named_columns = set()
+  for input, cols in fields.items():
+    if input not in tables:
+      return ValueError(f'invalid input {input}')
+    if isinstance(cols, list):
+      for col in cols:
+        if col in named_columns:
+          return ValueError(
+            f'same field name {col} specified more than once')
+        output_fields.append(f'{input}.{col} AS {col}')
+        named_columns.add(col)
+    elif isinstance(cols, dict):
+      for k, v in cols.items():
+        if k in named_columns:
+          return ValueError(
+            f'same field name {k} specified more than once')
+        output_fields.append(f'{input}.{v} AS {k}')
+        named_columns.add(k)
+    else:
+      return ValueError(f'invalid entry for fields')
+  for table in tables:
+    if table not in fields.keys():
+      output_fields.append(f'{table}.*')
+  return output_fields
+
+
+@beam.ptransform.ptransform_fn
+def _SqlJoinTransform(pcoll, sql_transform_constructor, type,
+                      equalities, fields=None):
+  _validate_type(type, str)
+  _validate_equalities(equalities, pcoll)
+
+  tables = list(pcoll)
+  outer = []
+  if isinstance(type, dict):
+    outer = type['outer']
+  base_table = tables[0]
+  conditioned = {base_table}
+
+  def generate_join_type(left, right):
+    if left in outer and right in outer:
+      return 'FULL'
+    if left in outer:
+      return 'LEFT'
+    if right in outer:
+      return 'RIGHT'
+    if not outer:
+      return type.upper()
+    return 'INNER'
+
+  prev_table = tables[0]
+  join_conditions = {}
+  for i in range(1, len(tables)):
+    curr_table = tables[i]
+    join_type = generate_join_type(prev_table, curr_table)
+    join_conditions[curr_table] = f' {join_type} JOIN {curr_table}'
+    prev_table = curr_table
+
+  for equality in equalities:
+    keys = list(equality)
+    left, right = keys[0], keys[1]
+    # TODO(titodo) - shorten this and reduce repitition
+    if left in conditioned and right in conditioned:
+      t = tables[max(tables.index(left), tables.index(right))]
```

Comment:
How do we know that the largest table (by index in tables) is the right one to add? E.g. what if I had tables A, B and C, but my equalities were `A.col1 = C.col1`, `A.col2 = B.col2`, and then `C.col3 = B.col3`?

Reply:
OK, that makes sense.